### PR TITLE
Return the logits 

### PR DESCRIPTION
**Pull request by ALGW71**
Flag that allows users to return the logits for a per-position scoring.

Addresses RC that asks can users see which positions are not confidently scored by the model.

### CODE REVIEW

**Comment by ALGW71** [comment]
@benjaminhwilliams you were correct - cannot see a clear speed difference over multiple runs.

File: src/anarcii/inference/model_runner.py [hunk 1]
```diff
@@ -270,17 +282,31 @@ def _predict_numbering(self, dl):
...
+                                {
+                                    "numbering": None,
+                                    "chain_type": "F",
+                                    "score": normalized_score,
+                                    "query_start": None,
+                                    "query_end": None,
+                                    "error": error_msg or "Score less than cut off.",
+                                    "scheme": "imgt",
+                                }
+                            )
```
Comment (benjaminhwilliams) [inline src/anarcii/inference/model_runner.py]: I notice that you are choosing not to leave a placeholder value for the item `logits` in the results dictionary.  Would it be sensible to do so?  That way, downstream tools that ingest the results dictionary could be assured of knowing the full set of keys in either scenario.  And it would allow you to simplify this block:
```suggestion
                        numbering.append(
                            {
                                "numbering": None,
                                "chain_type": "F",
                                "score": normalized_score,
                                "query_start": None,
                                "query_end": None,
                                "error": error_msg or "Score less than cut off.",
                                "scheme": "imgt",
                                "logits": all_scores if self.return_logits else None,
                            }
                        )
```

Comment (benjaminhwilliams) [reply src/anarcii/inference/model_runner.py]: If you prefer not to include a `NoneType` placeholder for the `logits` item, you could still simplify this block, to avoid duplication:
```suggestion
                        result =  {
                            "numbering": None,
                            "chain_type": "F",
                            "score": normalized_score,
                            "query_start": None,
                            "query_end": None,
                            "error": error_msg or "Score less than cut off.",
                            "scheme": "imgt",
                        }
                        if self.return_logits:
                            result["logits"] = all_scores
                        
                        numbering.append(result)
```

File: src/anarcii/inference/model_runner.py [hunk 2]
```diff
@@ -366,18 +392,33 @@ def _predict_numbering(self, dl):
...
+                                            "numbering": None,
+                                            "chain_type": "F",
+                                            "score": normalized_score,
+                                            "query_start": None,
+                                            "query_end": None,
+                                            "error": "Could not apply numbering: "
+                                            f"{captured_error}",
+                                            "scheme": "imgt",
+                                        }
+                                    )
```
Comment (benjaminhwilliams) [inline src/anarcii/inference/model_runner.py]: See comment above.  If you choose always to have a `logits` item in the results dictionary, this could be simplified to
```suggestion
                                numbering.append(
                                    {
                                        "numbering": None,
                                        "chain_type": "F",
                                        "score": normalized_score,
                                        "query_start": None,
                                        "query_end": None,
                                        "error": f"Could not apply numbering: {e}",
                                        "scheme": "imgt",
                                        "logits": all_scores if self.return_logits else None,
                                    }
                                )
```

Alternatively, if you don't want to keep a `NoneType` placeholder for `"logits"`, you could do
```suggestion
                                results = {
                                    "numbering": None,
                                    "chain_type": "F",
                                    "score": normalized_score,
                                    "query_start": None,
                                    "query_end": None,
                                    "error": f"Could not apply numbering: {e}",
                                    "scheme": "imgt",
                                }
                                if self.return_logits:
                                    results["logits"] = all_scores
                                
                                numbering.append(results)
```

Note that in both cases, line 394 is redundant because `f"{e}"` implies `str(e)`:https://github.com/oxpig/ANARCII/blob/1e930e64d906490f00f3d6f19d6f7748a2f5376e/src/anarcii/inference/model_runner.py#L393-L394

File: src/anarcii/inference/model_runner.py [hunk 3]
```diff
@@ -391,18 +432,33 @@ def _predict_numbering(self, dl):
...
```
Comment (benjaminhwilliams) [inline src/anarcii/inference/model_runner.py]: See comment above.  If you choose always to have a `logits` item in the results dictionary, this could be simplified to
```suggestion
                                numbering.append(
                                    {
                                        "numbering": None,
                                        "chain_type": "F",
                                        "score": normalized_score,
                                        "query_start": None,
                                        "query_end": None,
                                        "error": f"Could not apply numbering: {e}",
                                        "scheme": "imgt",
                                        "logits": all_scores if self.return_logits else None,
                                    }
                                )
```

Alternatively, if you don't want to keep a `NoneType` placeholder for `"logits"`, you could do
```suggestion
                                results = {
                                    "numbering": None,
                                    "chain_type": "F",
                                    "score": normalized_score,
                                    "query_start": None,
                                    "query_end": None,
                                    "error": f"Could not apply numbering: {e}",
                                    "scheme": "imgt",
                                }
                                if self.return_logits:
                                    results["logits"] = all_scores
                                
                                numbering.append(results)
```
Note that in both cases, line 434 is redundant becasue `f"{e}"` implies `str(e)`:https://github.com/oxpig/ANARCII/blob/1e930e64d906490f00f3d6f19d6f7748a2f5376e/src/anarcii/inference/model_runner.py#L433-L434

File: src/anarcii/inference/model_runner.py [hunk 4]
```diff
@@ -535,17 +605,33 @@ def _predict_numbering(self, dl):
...
+                                    "numbering": None,
+                                    "chain_type": "F",
+                                    "score": normalized_score,
+                                    "query_start": None,
+                                    "query_end": None,
+                                    "error": "Could not apply numbering: "
+                                    f"{captured_error}",
+                                    "scheme": "imgt",
+                                }
+                            )
```
Comment (benjaminhwilliams) [inline src/anarcii/inference/model_runner.py]: See comment above.  If you choose always to have a `logits` item in the results dictionary, this could be simplified to
```suggestion
                        numbering.append(
                            {
                                "numbering": None,
                                "chain_type": "F",
                                "score": normalized_score,
                                "query_start": None,
                                "query_end": None,
                                "error": f"Could not apply numbering: {e}",
                                "scheme": "imgt",
                                "logits": all_scores if self.return_logits else None,
                            }
                        )
```

Alternatively, if you don't want to keep a `NoneType` placeholder for `"logits"`, you could do
```suggestion
                        results = {
                            "numbering": None,
                            "chain_type": "F",
                            "score": normalized_score,
                            "query_start": None,
                            "query_end": None,
                            "error": f"Could not apply numbering: {e}",
                            "scheme": "imgt",
                        }
                        if self.return_logits:
                            results["logits"] = all_scores

                        numbering.append(results)
```
Note that in both cases, line 607 is redundant becasue `f"{e}"` implies `str(e)`:https://github.com/oxpig/ANARCII/blob/1e930e64d906490f00f3d6f19d6f7748a2f5376e/src/anarcii/inference/model_runner.py#L607

File: src/anarcii/inference/model_runner.py [hunk 5]
```diff
@@ -427,17 +483,31 @@ def _predict_numbering(self, dl):
...
+                                {
+                                    "numbering": None,
+                                    "chain_type": "F",
+                                    "score": normalized_score,
+                                    "query_start": None,
+                                    "query_end": None,
+                                    "error": "Model predicted duplicate numbers",
+                                    "scheme": "imgt",
+                                }
+                            )
```
Comment (benjaminhwilliams) [inline src/anarcii/inference/model_runner.py]: See comment above.  If you choose always to have a `logits` item in the results dictionary, this could be simplified to
```suggestion
                        numbering.append(
                            {
                                "numbering": None,
                                "chain_type": "F",
                                "score": normalized_score,
                                "query_start": None,
                                "query_end": None,
                                "error": "Model predicted duplicate numbers",
                                "scheme": "imgt",
                                "logits": all_scores if self.return_logits else None,
                            }
                        )
```

Alternatively, if you don't want to keep a `NoneType` placeholder for `"logits"`, you could do
```suggestion
                        results = {
                            "numbering": None,
                            "chain_type": "F",
                            "score": normalized_score,
                            "query_start": None,
                            "query_end": None,
                            "error": "Model predicted duplicate numbers",
                            "scheme": "imgt",
                        }
                        if self.return_logits:
                            results["logits"] = all_scores

                        numbering.append(results)
```

File: src/anarcii/inference/model_runner.py [hunk 6]
```diff
@@ -589,16 +675,29 @@ def _predict_numbering(self, dl):
...
+                            {
+                                "numbering": list(zip(nums, residues)),
+                                "chain_type": str(pred_tokens[batch_no, 1]),
+                                "score": normalized_score,
+                                "query_start": start_index,
+                                "query_end": end_index,
+                                "error": None,
+                                "scheme": "imgt",
+                            }
+                        )
```
Comment (benjaminhwilliams) [inline src/anarcii/inference/model_runner.py]: See comment above.  If you choose always to have a `logits` item in the results dictionary, this could be simplified to
```suggestion
                    numbering.append(
                        {
                            "numbering": list(zip(nums, residues)),
                            "chain_type": str(pred_tokens[batch_no, 1]),
                            "score": normalized_score,
                            "query_start": start_index,
                            "query_end": end_index,
                            "error": None,
                            "scheme": "imgt",
                            "logits": all_scores if self.return_logits else None,
                        }
                    )
```

Alternatively, if you don't want to keep a `NoneType` placeholder for `"logits"`, you could do
```suggestion
                    results = {
                        "numbering": list(zip(nums, residues)),
                        "chain_type": str(pred_tokens[batch_no, 1]),
                        "score": normalized_score,
                        "query_start": start_index,
                        "query_end": end_index,
                        "error": None,
                        "scheme": "imgt",
                    }
                    if self.return_logits:
                        results["logits"] = all_scores

                    numbering.append(results)
```

**Review by benjaminhwilliams — CHANGES_REQUESTED** [review]
There are a few more opportunities to cull duplicated code from this, which would help reduce the opportunity for mistakes to creep in over time.  See the suggestions for options, or adapt as you see fit.  Broadly speaking though, the functionality looks good.
